src/spectrum.rs

```diff
@@ -0,0 +1,397 @@
+        (flt(0.000058706520),flt(0.000021200000),flt(0.000000000000)),
+        (flt(0.000041509940),flt(0.000014990000),flt(0.000000000000)),
+        (flt(0.000029353260),flt(0.000010600000),flt(0.000000000000)),
+        (flt(0.000020673830),flt(0.000007465700),flt(0.000000000000)),
+        (flt(0.000014559770),flt(0.000005257800),flt(0.000000000000)),
+        (flt(0.000010253980),flt(0.000003702900),flt(0.000000000000)),
+        (flt(0.000007221456),flt(0.000002607800),flt(0.000000000000)),
+        (flt(0.000005085868),flt(0.000001836600),flt(0.000000000000)),
+        (flt(0.000003581652),flt(0.000001293400),flt(0.000000000000)),
+        (flt(0.000002522525),flt(0.000000910930),flt(0.000000000000)),
+        (flt(0.000001776509),flt(0.000000641530),flt(0.000000000000)),
+        (flt(0.000001251141),flt(0.000000451810),flt(0.000000000000)),
+    ]
+}
+
+#[inline(always)]
+fn lerp<T: Float>(a: T, b: T, t: T) -> T {
+    a + (b - a) * t
+}
+
+#[derive(Clone, Copy, Eq, PartialEq)]
+pub struct Sample<T: Float>(OrderedFloat<f32>, T);
+
+impl<T: Float> Sample<T> {
+
+    /// Create a new Sample.
+    ///
+    /// # Panics
+    /// If the intensity is less than zero.
+    pub fn new(wavelength: f32, intensity: T) -> Sample<T> {
+        assert!(intensity >= T::zero(), "Intensity must be >= 0.");
+        Sample(OrderedFloat(wavelength), intensity)
+    }
+
+    #[inline(always)]
+    pub fn wavelength(&self) -> f32 {
+        self.0.into_inner()
+    }
+
+    #[inline(always)]
+    pub fn intensity(&self) -> T {
+        self.1
+    }
+}
+
+/// Spectrum representing the relative intensity of
+/// wavelengths from 360 nm to 830 nm.
+pub struct Spectrum<T: Float> {
+    data: [T; N_SAMPLES],
+}
+
+impl<T: Float> Clone for Spectrum<T> {
+    fn clone(&self) -> Spectrum<T> {
+        let mut data: [T; N_SAMPLES] = [T::zero(); N_SAMPLES];
+        data.copy_from_slice(&self.data);
+        Spectrum {
+            data: data,
+        }
+    }
+}
+
+impl<T> Spectrum<T>
+    where T: Float + Zero + FromPrimitive
+{
+    /// Create a Spectrum from an array of spectral intensity
+    /// values assumed to be in the range 360 nm to 830 nm.
+    ///
+    /// The intensity values must be greater than or equal to
+    /// zero and not NaN.
+    ///
+    /// # Panics
+    /// If an intensity value is less than zero.
+    pub fn new(data: [T; N_SAMPLES]) -> Spectrum<T> {
+        assert!(data.iter().all(|&intensity| intensity >= T::zero()));
+        Spectrum { data: data }
+    }
+
+    /// Create a `Spectrum` from a slice of (`f32`, `Float`) tuples
+    /// representing a (wavelength, intensity) mapping.
+    /// 
+    /// The more data points there are the more accurate the 
+    /// `Spectrum`'s internal representation will be.
+    ///
+    /// The intensity values must be greater than or equal to
+    /// zero and not NaN.
+    ///
+    pub fn from_samples(data: &[Sample<T>]) -> Spectrum<T> {
```

Comment:
This could take `&[S]` where `S: Into<Sample<T>>`. And `Sample<T>` could implement `From<(f32, T)>`.